.github/workflows/build_and_test.yml

```diff
@@ -0,0 +1,256 @@
+name: Build and test
+
+on:
+  push:
+    branches:
+      - master
+      - "*-rc"
+  pull_request:
+    types:
+      - opened
+      - synchronize
+      - reopened
+    branches:
+      - "**"
+
+jobs:
+  verify_files:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup
+        run: |
+          sudo apt-get update -y && sudo apt-get install -y gnupg2
+
+      - name: Verify files
+        run: |
+          curl -sSL https://secchannel.rsk.co/SUPPORT.asc | gpg --import -
+          gpg --verify SHA256SUMS.asc && sha256sum --check SHA256SUMS.asc
+
+  clone_rskj_repo:
+    needs: verify_files
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout RSKj repo
+        uses: actions/checkout@v4
+        with:
+          repository: rsksmart/rskj
+          token: ${{ secrets.GITHUB_TOKEN }}
+          path: rskj
+          fetch-depth: 0
+      
+      - name: Determine branch to checkout for PR
+        if: github.event_name == 'pull_request'
+        working-directory: rskj
+        env:
+          PR_BRANCH: ${{ github.head_ref }}
+        run: |
+          IS_RSKJ_BRANCH=`git ls-remote --heads origin $PR_BRANCH`
+          if test -n "${IS_RSKJ_BRANCH}"; then
+            echo "Found matching branch name in RSKj repo"
+            CHECKOUT_REF="$PR_BRANCH"
+          else
+            echo "Using master for RSKj"
+            CHECKOUT_REF="master"
+          fi
+          echo "CHECKOUT_REF=$CHECKOUT_REF" >> $GITHUB_ENV
+          
+      - name: Determine branch to checkout for push 
+        if: github.event_name != 'pull_request'
+        working-directory: rskj
+        env:
+          POW_REF: ${{ github.ref }}
+        run: |
+          IS_RSKJ_REF=`git ls-remote --refs $POW_REF`
+          if test -n "${IS_RSKJ_REF}"; then
+            echo "Found matching ref in RSKj"
+            CHECKOUT_REF="$POW_REF"
+          else
+            echo "Using master for RSKj"
+            CHECKOUT_REF="master"
+          fi
+          echo "CHECKOUT_REF=$CHECKOUT_REF" >> $GITHUB_ENV
+
+      - name: Check out appropriate rskj reference
+        working-directory: rskj
+        run: |
+          git switch "${{ env.CHECKOUT_REF }}"
```

Comment:
Manage this as `POW_REF` was managed in the previous step.

Reply:
The workflow  already aligns well with how `POW_REF` was handled in the previous step. Both `CHECKOUT_REF` and `POW_REF` are set based on whether it's a pull request or a direct push event. We check if the branch or reference exists in the repository using git ls-remote and default to master if there's no match found. This approach ensures consistency and clarity throughout the workflow. It makes sure we're always checking out the right branch or reference depending on how the GitHub action was triggered.

If there's anything specific you think needs adjusting or any other concerns, just let me know!

Reply:
With the current implementation, I can create a branch called `main";env;echo"` which will render the command as

```
git switch "main";env;echo""
```

showing all the environments variable in the workflow. I can change `env` for a `curl` and further leak information.

If you replace it with

```
env:
  CHECKOUT_REF: ${{ env.CHECKOUT_REF }}
run: |
  git switch "$CHECKOUT_REF"
```

the issue is gone, because it's the shell doing the replacement and not the template engine of GHA.

Reply:
Bonus comment: `CHECKOUT_REF` is defined in the environment only in the case that the event isn't a pull request (the condition of the previous step), yet this step is always ran.

Reply:
Thanks for the clarification! I've implemented the solution you proposed